### PR TITLE
Enable filtering ZAP extensions in addons pop-up

### LINKED ISSUE #2334: Enable searching in ZAP Addons Pop-up
Hi, 

as the number of ZAP extensions is growing, the addons pop-up should have a search box in order to be user-friendly.

I can work on this feature. Will need some guidance where to start also ;)

### PR DESCRIPTION
**Pull request by yaziza**
Add a Filter text Field to the installed and
uninstalled(Marketplace) addons panels

Fix #2334 - Enable searching in ZAP Addons Pop-up

### CODE REVIEW

File: src/lang/Messages.properties [hunk 1]
```diff
@@ -586,6 +586,7 @@ To install it either:<ul>\
 cfu.label.cfuonstart = Check for updates on start?
 cfu.label.dir.border   = Add-on Directories
 cfu.label.addons.border = Add-ons
+cfu.label.addons.filter = Filter
```
Comment (thc202) [inline src/lang/Messages.properties]: Maybe add a colon at the end?

File: src/org/zaproxy/zap/extension/autoupdate/ManageAddOnsDialog.java [hunk 2]
```diff
@@ -254,14 +259,17 @@ private JPanel getInstalledAddOnsPanel() {
 							java.awt.Color.black));
 
 			getInstalledAddOnsTable();
-			
+
```
Comment (thc202) [inline src/org/zaproxy/zap/extension/autoupdate/ManageAddOnsDialog.java]: Unrelated indentation changes should not be done in the same commit. [1]

[1] https://github.com/zaproxy/zaproxy/wiki/DevGuidelines

File: src/org/zaproxy/zap/extension/autoupdate/ManageAddOnsDialog.java [hunk 3]
```diff
@@ -254,14 +259,17 @@ private JPanel getInstalledAddOnsPanel() {
 							java.awt.Color.black));
 
 			getInstalledAddOnsTable();
-			
+
 			JScrollPane scrollPane = new JScrollPane();
 			scrollPane.setHorizontalScrollBarPolicy(javax.swing.JScrollPane.HORIZONTAL_SCROLLBAR_NEVER);
 			scrollPane.setViewportView(getInstalledAddOnsTable());
 
+			JPanel searchPanel = getSearchPanel(getInstalledAddOnsTable());
```
Comment (thc202) [inline src/org/zaproxy/zap/extension/autoupdate/ManageAddOnsDialog.java]: `filterPanel` instead? See following comment(s) about the rename.

File: src/org/zaproxy/zap/extension/autoupdate/ManageAddOnsDialog.java [hunk 4]
```diff
@@ -254,14 +259,17 @@ private JPanel getInstalledAddOnsPanel() {
...
 			scrollPane.setHorizontalScrollBarPolicy(javax.swing.JScrollPane.HORIZONTAL_SCROLLBAR_NEVER);
 			scrollPane.setViewportView(getInstalledAddOnsTable());
 
+			JPanel searchPanel = getSearchPanel(getInstalledAddOnsTable());
+
 			int row = 0;
+			installedAddOnsPanel.add(searchPanel, LayoutHelper.getGBC(0, row++, 5, 0.0D));
 			installedAddOnsPanel.add(scrollPane, LayoutHelper.getGBC(0, row++, 5, 1.0D, 1.0D));
-			installedAddOnsPanel.add(new JLabel(""), LayoutHelper.getGBC(0, row, 1, 1.0D));
+			installedAddOnsPanel.add(new JLabel(""), LayoutHelper.getGBC(0, row++, 1, 1.0D));
```
Comment (thc202) [inline src/org/zaproxy/zap/extension/autoupdate/ManageAddOnsDialog.java]: This change does not seem to be needed.

File: src/org/zaproxy/zap/extension/autoupdate/ManageAddOnsDialog.java [hunk 5]
```diff
@@ -291,7 +299,10 @@ private JPanel getUninstalledAddOnsPanel() {
 				getMarketPlaceScrollPane().setViewportView(getUninstalledAddOnsTable());
 			}
 
+			JPanel searchPanel = getSearchPanel(getUninstalledAddOnsTable());
```
Comment (thc202) [inline src/org/zaproxy/zap/extension/autoupdate/ManageAddOnsDialog.java]: Same comment regarding the rename.
The filter panel should not be visible until the table is shown, it's odd seeing the filter panel alone with the "Check for Updates" button.

File: src/org/zaproxy/zap/extension/autoupdate/ManageAddOnsDialog.java [hunk 6]
```diff
@@ -301,7 +312,50 @@ private JPanel getUninstalledAddOnsPanel() {
 		}
 		return uninstalledAddOnsPanel;
 	}
-	
+
+	private JPanel getSearchPanel(final JXTable table) {
```
Comment (thc202) [inline src/org/zaproxy/zap/extension/autoupdate/ManageAddOnsDialog.java]: Probably better to use `create` instead of `get`, also `Filter` instead of `Search`, it describes better want the method does.
The "static" modifier should be added to the method.

File: src/org/zaproxy/zap/extension/autoupdate/ManageAddOnsDialog.java [hunk 7]
```diff
@@ -301,7 +312,50 @@ private JPanel getUninstalledAddOnsPanel() {
 		}
 		return uninstalledAddOnsPanel;
 	}
-	
+
+	private JPanel getSearchPanel(final JXTable table) {
+		JPanel searchPanel = new JPanel();
```
Comment (thc202) [inline src/org/zaproxy/zap/extension/autoupdate/ManageAddOnsDialog.java]: As the above comment, this should be renamed to `filterPanel`.

File: src/org/zaproxy/zap/extension/autoupdate/ManageAddOnsDialog.java [hunk 8]
```diff
@@ -301,7 +312,50 @@ private JPanel getUninstalledAddOnsPanel() {
 		}
 		return uninstalledAddOnsPanel;
 	}
-	
+
+	private JPanel getSearchPanel(final JXTable table) {
+		JPanel searchPanel = new JPanel();
+		searchPanel.setLayout(new GridBagLayout());
+
+		JLabel lookupLabel = new JLabel(Constant.messages.getString("cfu.label.addons.filter"));
```
Comment (thc202) [inline src/org/zaproxy/zap/extension/autoupdate/ManageAddOnsDialog.java]: Ditto but use `filter` instead of `lookup`. (Same for text field.)
The text field should be set as the component of the label (`JLabel.setLabelFor(Component)`).
Consider setting a tool tip for both the label and text field (or at least one of them) indicating that the field supports regular expressions.

File: src/org/zaproxy/zap/extension/autoupdate/ManageAddOnsDialog.java [hunk 9]
```diff
@@ -301,7 +312,50 @@ private JPanel getUninstalledAddOnsPanel() {
...
+				updateFilter();
+			}
+
+			@Override
+			public void changedUpdate(DocumentEvent e) {
+				updateFilter();
+			}
+
+			public void updateFilter() {
+				try {
```
Comment (thc202) [inline src/org/zaproxy/zap/extension/autoupdate/ManageAddOnsDialog.java]: The try-catch statement should be moved to the else block.

File: src/org/zaproxy/zap/extension/autoupdate/ManageAddOnsDialog.java [hunk 10]
```diff
@@ -301,7 +312,50 @@ private JPanel getUninstalledAddOnsPanel() {
...
+					String lookupFieldText = lookupField.getText();
+					if (lookupFieldText.length() == 0) {
```
Comment (thc202) [inline src/org/zaproxy/zap/extension/autoupdate/ManageAddOnsDialog.java]: Consider using `String.isEmpty()` instead.

File: src/org/zaproxy/zap/extension/autoupdate/ManageAddOnsDialog.java [hunk 11]
```diff
@@ -301,7 +312,50 @@ private JPanel getUninstalledAddOnsPanel() {
...
+						table.setRowFilter(null);
+					} else {
+						table.setRowFilter(RowFilter.regexFilter("(?i)" + lookupFieldText));
+					}
+				} catch (PatternSyntaxException e) {
+					logger.error("Incorrect pattern syntax " + e);
```
Comment (thc202) [inline src/org/zaproxy/zap/extension/autoupdate/ManageAddOnsDialog.java]: It should not be logged as error, this is expected to happen, not a bug or issue in ZAP. If you really want to log the exception better to use debug level instead.

It would be great to give some indication that the regex is wrong, maybe with a visual change in the text field?

**Comment by thc202** [comment]
There's a typo in the commit message "Fieled".

The branch needs to be rebased on develop branch to fix the conflicts.

**Comment by kingthorin** [comment]
@yaziza more coming?

**Comment by yaziza** [comment]
@kingthorin 

yes :)

i will update the patch tomorrow, sorry was busy last weeks and without internet access :/

**Comment by kingthorin** [comment]
No problem :smiley:

**Comment by yaziza** [comment]
Patch updated !

File: src/org/zaproxy/zap/extension/autoupdate/ManageAddOnsDialog.java [hunk 12]
```diff
@@ -286,15 +296,21 @@ private JPanel getUninstalledAddOnsPanel() {
...
 				// Not checked yet
 				getUninstalledAddOnsTable();	// To initialise the table and model
 				getMarketPlaceScrollPane().setViewportView(getRetrievePanel());
+				uninstalledAddOnsFilterPanel.setVisible(false);
 			} else {
 				getMarketPlaceScrollPane().setViewportView(getUninstalledAddOnsTable());
+				uninstalledAddOnsFilterPanel.setVisible(true);
 			}
 
+
```
Comment (thc202) [inline src/org/zaproxy/zap/extension/autoupdate/ManageAddOnsDialog.java]: This extra empty line can be removed.

File: src/org/zaproxy/zap/extension/autoupdate/ManageAddOnsDialog.java [hunk 13]
```diff
@@ -304,7 +320,56 @@ private JPanel getUninstalledAddOnsPanel() {
...
+			public void updateFilter() {
+				String filterText = filterTextField.getText();
+				if (filterText.isEmpty()) {
+					table.setRowFilter(null);
+				} else {
+					try {
+						table.setRowFilter(RowFilter.regexFilter("(?i)" + filterText));
+					} catch (PatternSyntaxException e) {
+						String invalidPattern = Constant.messages.getString("cfu.label.addons.filter.invalidpattern");
+						filterTextField.setText(invalidPattern + filterText);
```
Comment (thc202) [inline src/org/zaproxy/zap/extension/autoupdate/ManageAddOnsDialog.java]: This is leading to an exception:
```
65145 [AWT-EventQueue-0] ERROR org.zaproxy.zap.ZAP$UncaughtExceptionLogger  - Exception in thread "AWT-EventQueue-0"
java.lang.IllegalStateException: Attempt to mutate in notification
	at javax.swing.text.AbstractDocument.writeLock(AbstractDocument.java:1338)
	at javax.swing.text.AbstractDocument.replace(AbstractDocument.java:659)
	at javax.swing.text.JTextComponent.setText(JTextComponent.java:1718)
	at org.zaproxy.zap.extension.autoupdate.ManageAddOnsDialog$1.updateFilter(ManageAddOnsDialog.java:365)
	at org.zaproxy.zap.extension.autoupdate.ManageAddOnsDialog$1.insertUpdate(ManageAddOnsDialog.java:343)
```
I was thinking on a change of back/foreground colours, but an icon or even a message somewhere (under the text box?) is probably better.
Changing the text of the text box would also cause usability problems, since the user would have to delete the text set to fix the issues.

File: src/lang/Messages.properties [hunk 14]
```diff
@@ -581,6 +581,9 @@ To install it either:<ul>\
 cfu.label.cfuonstart = Check for updates on start?
 cfu.label.dir.border   = Add-on Directories
 cfu.label.addons.border = Add-ons
+cfu.label.addons.filter = Filter:
+cfu.label.addons.filter.tooltip = The add-ons filtering system supports regular expression searches.
```
Comment (thc202) [inline src/lang/Messages.properties]: I would remove the "searches" and change to plural "expressions".

**Comment by kingthorin** [comment]
@yaziza looks close, could you address these last few comments?

File: src/lang/Messages.properties (hunk 14 above)
Comment (yaziza) [inline src/lang/Messages.properties]: done

**Comment by yaziza** [comment]
@kingthorin, patch updated.

Now the foreground color will be changed to red in case the user types an invalid regexp. 

Sorry for the delay !

File: src/lang/Messages.properties (hunk 14 above)
Comment (kingthorin) [inline src/lang/Messages.properties]: Something's still off here.

**Comment by yaziza** [comment]
@kingthorin , patch updated.

Thanx for your feedback !

File: src/org/zaproxy/zap/extension/autoupdate/ManageAddOnsDialog.java [hunk 15]
```diff
@@ -304,7 +321,56 @@ private JPanel getUninstalledAddOnsPanel() {
 		}
 		return uninstalledAddOnsPanel;
 	}
-	
+
```
Comment (thc202) [inline src/org/zaproxy/zap/extension/autoupdate/ManageAddOnsDialog.java]: This (minor) indentation change could be reverted.

File: src/org/zaproxy/zap/extension/autoupdate/ManageAddOnsDialog.java [hunk 16]
```diff
@@ -304,7 +321,56 @@ private JPanel getUninstalledAddOnsPanel() {
...
+			@Override
+			public void changedUpdate(DocumentEvent e) {
+				updateFilter();
+			}
+
+			public void updateFilter() {
+				String filterText = filterTextField.getText();
+				if (filterText.isEmpty()) {
+					table.setRowFilter(null);
```
Comment (thc202) [inline src/org/zaproxy/zap/extension/autoupdate/ManageAddOnsDialog.java]: The foreground colour should also be reset here (otherwise if the whole "wrong regex" is removed the cursor would still be shown as red).

Comment (yaziza) [inline src/org/zaproxy/zap/extension/autoupdate/ManageAddOnsDialog.java]: done

**Comment by kingthorin** [comment]
@yaziza thanks!

**Comment by thc202** [comment]
Looks good to me. Thank you!
